test/builder/patch.js

```diff
@@ -0,0 +1,326 @@
+    }, {});
+  }
+
+  getInsertionPoint() {
+    return this.buffer.getEndPosition();
+  }
+
+  getLayer(markerLayerName) {
+    const layer = this.layers[markerLayerName];
+    if (!layer) {
+      throw new Error(`invalid marker layer name: ${markerLayerName}`);
+    }
+    return layer;
+  }
+
+  appendMarked(markerLayerName, lines) {
+    const startPosition = this.buffer.getEndPosition();
+    const layer = this.getLayer(markerLayerName);
+    this.buffer.append(lines.join('\n'));
+    const marker = layer.markRange([startPosition, this.buffer.getEndPosition()], {exclusive: true});
+    this.buffer.append('\n');
+    return marker;
+  }
+
+  markFrom(markerLayerName, startPosition) {
+    const endPosition = this.buffer.getEndPosition().translate([-1, Infinity]);
+    const layer = this.getLayer(markerLayerName);
+    return layer.markRange([startPosition, endPosition], {exclusive: true});
+  }
+
+  wrapReturn(object) {
+    return {
+      buffer: this.buffer,
+      layers: this.layers,
+      ...object,
+    };
+  }
+}
+
+class MultiFilePatchBuilder {
+  constructor(layeredBuffer = null) {
+    this.layeredBuffer = layeredBuffer;
+
+    this.filePatches = [];
+  }
+
+  addFilePatch(block = () => {}) {
+    const filePatch = new FilePatchBuilder(this.layeredBuffer);
+    block(filePatch);
+    this.filePatches.push(filePatch.build().filePatch);
+    return this;
+  }
+
+  build() {
+    return this.layeredBuffer.wrapReturn({
+      multiFilePatch: new MultiFilePatch({
+        buffer: this.layeredBuffer.buffer,
+        layers: this.layeredBuffer.layers,
+        filePatches: this.filePatches,
+      }),
+    });
+  }
+}
+
+class FilePatchBuilder {
+  constructor(layeredBuffer = null) {
+    this.layeredBuffer = layeredBuffer;
+
+    this.oldFile = new File({path: 'file', mode: '100644'});
+    this.newFile = null;
+
+    this.patchBuilder = new PatchBuilder(this.layeredBuffer);
+  }
+
+  setOldFile(block) {
+    const file = new FileBuilder();
+    block(file);
+    this.oldFile = file.build().file;
+    return this;
+  }
+
+  nullOldFile() {
+    this.oldFile = nullFile;
+    return this;
+  }
+
+  setNewFile(block) {
+    const file = new FileBuilder();
+    block(file);
+    this.newFile = file.build().file;
+    return this;
+  }
+
+  nullNewFile() {
+    this.newFile = nullFile;
+    return this;
+  }
+
+  status(...args) {
+    this.patchBuilder.status(...args);
+    return this;
+  }
+
+  addHunk(...args) {
+    this.patchBuilder.addHunk(...args);
+    return this;
+  }
+
+  empty() {
+    this.patchBuilder.empty();
+    return this;
+  }
+
+  build() {
+    const {patch} = this.patchBuilder.build();
+
+    if (this.newFile === null) {
+      this.newFile = this.oldFile.clone();
+    }
+
+    return this.layeredBuffer.wrapReturn({
+      filePatch: new FilePatch(this.oldFile, this.newFile, patch),
+    });
+  }
+}
+
+class FileBuilder {
+  constructor() {
+    this._path = 'file.txt';
+    this._mode = '100644';
```

Comment:
ugh, I can never remember what the file modes mean.  Is this worth documenting with a comment?

Reply:
🚨 I would prefer extracting them to constants, personally. Probably in a static property on the File class? I can handle it 👌